### ISSUE TITLE
Unexpected token with contextual lexer

### ISSUE THREAD

**Issue by tjume** [lark-parser/lark#517]
I am not sure whether this is a bug or just me not understanding how the contextual lexer is supposed to work. I have reduced my problem to the following code.

```python
import lark

parser = lark.Lark("""
start: (a "b" | "b" | d)+
a: "a"
d: "c" a "bb"
""", parser='lalr', lexer='contextual')

parser.parse('abb')
```

Running this gives the following exception.

```
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/lark/parsers/lalr_parser.py", line 62, in get_action
    return states[state][token.type]
KeyError: 'BB'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./contextual-parser-test", line 11, in <module>
    parser.parse('abb')
  File "/usr/lib/python3.8/site-packages/lark/lark.py", line 311, in parse
    return self.parser.parse(text, start=start)
  File "/usr/lib/python3.8/site-packages/lark/parser_frontends.py", line 89, in parse
    return self._parse(token_stream, start, *[sps] if sps is not NotImplemented else [])
  File "/usr/lib/python3.8/site-packages/lark/parser_frontends.py", line 54, in _parse
    return self.parser.parse(input, start, *args)
  File "/usr/lib/python3.8/site-packages/lark/parsers/lalr_parser.py", line 36, in parse
    return self.parser.parse(*args)
  File "/usr/lib/python3.8/site-packages/lark/parsers/lalr_parser.py", line 86, in parse
    action, arg = get_action(token)
  File "/usr/lib/python3.8/site-packages/lark/parsers/lalr_parser.py", line 65, in get_action
    raise UnexpectedToken(token, expected, state=state)
lark.exceptions.UnexpectedToken: Unexpected token Token(BB, 'bb') at line 1, column 2.
Expected one of: 
	* B
```

So it seems like the `bb` in the input `abb` is transformed by the lexer into a `BB` token. My understanding of the contextual lexer was that by communicating with the parser, it only tries tokens which are actually possible at the current position, meaning that the `BB` token would only be used if we have already parsed the `"c" a` before it. Since there is no `"c"` in the string, this can not happen and therefore I thought that `BB` would never be used.

**Comment by erezsh** [comment]
Hi,

First of all, I would just like to say, great bug report!

As for your question: Yes, it is the intended behavior. I will try to explain why.

Here is a simpler example of your issue:
```python
import lark
parser = lark.Lark("""
start: a "b" "b"
       | "c" a "bb"
a: "a"
""", parser='lalr', lexer='contextual')
parser.parse('abb')
```

Lark will fail with the same error.

The reason is that when it computes the states and lookaheads, the state `<a: A * >`, which means that the letter 'a' had just been seen, and the rule `a` is about to be reduced, can return to either derivation of `start`. While the LALR algorithm knows that these are two separate states, it combines them on purpose for the parser, so it will have a smaller number of states.

It might be possible to fix, but it certainly won't be easy. It can try to split the offending state just before creating the parse table, according to information from the lexer, and fix all the pointers to it.

Or LALR(2).

Meanwhile, either use priority to always choose one over the other, or try to refactor your grammar so that this ambiguity doesn't occur immediately after the same reduce.

**Comment by tjume** [comment]
Thanks for the explanation. That makes sense. I was also able to refactor my grammar so that this is no problem for me anymore.

**Comment by handras** [comment]
@tjume could you please show how to refactor the grammar? I am facing the same issue with unexpected tokens, although there is only one possible derivation.

**Comment by tjume** [comment]
@handras I think the general idea I used was to make my tokens longer, so that there is no ambiguity whether something should be lexed as `BB` or `B` `B` tokens but e.g. the lexer can only decide between a `BBC` and `B` `B` `D` tokens. So add whatever can follow the current problematic token to the token until ambiguity is removed.

That might lead to a bunch of refactoring if your problematic token is at the end of some rule. At least I think it did for me. Unfortunately I don't remember and haven't saved all the changes I did to my grammar and for your grammar the necessary changes might be different anyway.

**Comment by handras** [comment]
Thanks for your response. Unfortunately I can't change the grammar since it is defined by a standard. I was hoping there is a way to rewrite the grammar in a way so that this state merging does not happen.

**Comment by tjume** [comment]
I also meant to rewrite it, not to change the language it recognizes. You have to make it clear for the lexer in the problematic state which token it should use. This could either be done by adding priorities to the tokens, if it is clear that it should always try one token first. Or, if that does not work, then make the tokens longer so that your 1-token-lookahead is basically more powerful. And if that still doesn't work then think about whether your language can really be recognized by a LALR(1) parser or whether you need a different one.

**Comment by tjume** [comment]
Actually I might be wrong, it might also be possible to rewrite the grammar in some way without making the tokens longer or using priorities to solve the ambiguity. But it's been a while since I've worked on it and it would take me a while to figure that out. Sorry, I can't really help you more than that, I am uncertain about possible other solutions myself.

**Comment by erezsh** [comment]
@handras It depends on your grammar and usage of Lark. You shouldn't assume that you have the same problem and same solution as this issue, just because you're getting the same exception. Just like getting a KeyError or ValueError in Python doesn't mean the solutions are always the same.

**Comment by handras** [comment]
So here is a short extract based on the DBC grammar I am working on:

`grammar="""
%import common.CNAME
%import common.INT
%import common.WS
%ignore WS

start : signal* //group
signal : "SG_" signal_name multiplexer_indicator ";" 
signal_name : CNAME
!multiplexer_indicator : | "M" | "m" INT

//group: "GR_" CNAME ":" signal_name*

"""
parser = lark.Lark(grammar, parser='lalr', lexer='contextual')
print(parser.parse('SG_ motortemp m8 ; SG_ motorrev m8 ;').pretty())
`
In its current state it can parse the example string. 
However when the 'group' rule is introduced the lexer takes 'm8' as a CNAME token instead of 'm'.

My understanding is that the problem comes from the fact that the 'group' rule intruduces the possibility that a CNAME is followed by another CNAME. However it's not clear for me how the contextual lexer decides which token to try to match.

Prioritizing the "m" token is not working. If I increase the priority of matching "m" that introduces the problem that 'motortemp' will no longer be a CNAME token, but a single "m".

What I did as a workaround is that I defined a new terminal 'SWITCH.2 : /m\d+/'. So it has priority over CNAME which means there can't be signals named starting with this token. This is acceptable.

Furthermore there are other issues in the grammar where INT and DOUBLE tokens get confused.

Maybe this means that a LALR parser is not appropriate for this grammar. I do this parser for a Compilers course where we only discussed LL and LR parsers. I'm sure this grammar is LR parseable, however I am not aware of the limitations a LALR parser has.
